pkg/cluster/base.go

```diff
@@ -364,6 +374,86 @@ func (r *CreateClusterRequest) validateMainFields() error {
 	return nil
 }
 
+// constants for desiredSpotPrice validation
+const (
+	maxZonePrice     = "maxZonePrice"
+	desiredSpotPrice = "desiredSpotPrice"
+)
+
+func (r *CreateClusterRequest) validateAmazonSpotPrices() error {
+	var errs []error
+	if r.Properties.CreateClusterPKE != nil {
+		for _, np := range r.Properties.CreateClusterPKE.NodePools {
+			providerConfig := internalPke.NodePoolProviderConfigAmazon{}
+			if err := mapstructure.Decode(np.ProviderConfig, &providerConfig); err != nil {
+				return errors.WrapIff(err, "decoding nodepool %q config", np.Name)
+			}
+
+			// check if spotPrice is specified
+			if providerConfig.AutoScalingGroup.SpotPrice == "" {
+				continue
+			}
+
+			spotPriceMap, err := r.getSpotPriceMap(providerConfig.AutoScalingGroup.SpotPrice, cloud.Amazon, PKE, providerConfig.AutoScalingGroup.InstanceType)
+			if err != nil {
+				return err
+			}
+
+			for _, zone := range providerConfig.AutoScalingGroup.Zones {
+				if spotPriceMap[desiredSpotPrice] < spotPriceMap[string(zone)] {
+					errs = append(errs, pkgErrors.NewSpotPriceError(spotPriceMap[desiredSpotPrice], spotPriceMap[string(zone)], string(zone), providerConfig.AutoScalingGroup.InstanceType))
+				}
+			}
+
+			if errs == nil && spotPriceMap[desiredSpotPrice] < spotPriceMap[maxZonePrice] {
+				errs = append(errs, pkgErrors.NewSpotPriceError(spotPriceMap[desiredSpotPrice], spotPriceMap[maxZonePrice], r.Location, providerConfig.AutoScalingGroup.InstanceType))
+			}
+		}
+	} else {
+		for _, np := range r.Properties.CreateClusterEKS.NodePools {
+			// check if spotPrice is specified
+			if np.SpotPrice == "" {
+				continue
+			}
+
+			spotPriceMap, err := r.getSpotPriceMap(np.SpotPrice, cloud.Amazon, EKS, np.InstanceType)
+			if err != nil {
+				return err
+			}
+
+			if spotPriceMap[desiredSpotPrice] < spotPriceMap[maxZonePrice] {
+				errs = append(errs, pkgErrors.NewSpotPriceError(spotPriceMap[desiredSpotPrice], spotPriceMap[maxZonePrice], r.Location, np.InstanceType))
+			}
+		}
+	}
+
+	return errors.Combine(errs...)
+}
+
+func (r *CreateClusterRequest) getSpotPriceMap(spotPrice, cloud, service, instanceType string) (map[string]float64, error) {
+	spotPriceFloat, err := strconv.ParseFloat(spotPrice, 64)
+	if err != nil {
+		return nil, errors.WrapWithDetails(err, "could not parse spotPrice string to float64")
+	}
+
+	prices := make(map[string]float64)
+	prices[desiredSpotPrice] = spotPriceFloat
```

Comment:
**Question**: why are we adding the desired spot price to the map of spot prices here?

It is available from earlier, feels like unnecessary addition.
Otherwise the `maxZonePrice` could be replaced by the region name and then we wouldn't need globals, because the region is also available in case the request is region based thus we don't need extra keys for these things.
Also if we do use global extra keys they should be named appropriately (maxZonePrice as a global makes no sense to be set to the string `maxZonePrice`, it should be `maxZonePriceKey` or `maxZonePriceMapKey` to disambiguate, but I would only go for globals if I have to and this case it feels like we are only using these from 1 function here, and maybe where the map is processed, but there the necessary data such as desired spot price or the region in case of my proposal are already available, so no need for extra keys, region can be passed in to this function and the desired should not be put in the spot prices map, the spot prices map should contain the actual values, excluding the manual desired one IMO).